src/test/java/seedu/address/model/module/ModuleTest.java

```diff
@@ -0,0 +1,118 @@
+import seedu.address.testutil.TaskBuilder;
+import seedu.address.testutil.TypicalStudents;
+import seedu.address.testutil.TypicalTasks;
+
+class ModuleTest {
+
+    @Test
+    void isSameModule() {
+        // same object -> returns true
+        assertTrue(MODULE_1.isSameModule(MODULE_1));
+
+        // null -> returns false
+        assertFalse(MODULE_1.isSameModule(null));
+
+        // different module -> returns false
+        assertFalse(MODULE_1.isSameModule(MODULE_2));
+
+        // same name -> returns true
+        Module sameName = new ModuleBuilder().withName("CS2103").build();
+        assertTrue(MODULE_1.isSameModule(sameName));
+
+        // different name and tasks, same students -> returns false
+        Module differentName = new ModuleBuilder().withName("CS2105")
+                .withStudents(TypicalStudents.getTypicalStudents())
+                .withTasks(TypicalTasks.getTypicalTasksForModule("CS2105"))
+                .build();
+        assertFalse(MODULE_1.isSameModule(differentName));
+    }
+
+    @Test
+    void hasStudent() {
+        // initialised with the student -> returns true
+        assertTrue(MODULE_1.hasStudent(AMY));
+
+        // not initialised with the student -> returns false
+        assertFalse(MODULE_1.hasStudent(CHARLIE));
+    }
+
+    @Test
+    void hasTask() {
+        // initialised with the task -> returns true
+        Task task1 = new TaskBuilder()
+                .withModule("CS2103")
+                .withId("T1")
+                .withName("Assignment 1")
+                .withDeadline("2021-12-31").build();
+        assertTrue(MODULE_1.hasTask(task1));
+
+        // not initialised with the task -> returns false
+        Task task5 = new TaskBuilder()
+                .withModule("CS2103")
+                .withId("T5")
+                .withName("Task 5")
+                .withDeadline("2021-11-17 11:11").build();
+        assertFalse(MODULE_1.hasTask(task5));
+    }
+
+    @Test
+    void equals() {
+        // same object -> returns true
+        assertTrue(MODULE_1.equals(MODULE_1));
+
+        // not a Module object -> returns false
+        assertFalse(MODULE_1.equals(AMY));
+
+        // same attributes -> returns true
+        Module sameAttributes = new ModuleBuilder()
+                .withName("CS2103")
+                .build();
+        assertTrue(MODULE_1.equals(sameAttributes));
+
+        // null -> returns false
+        assertFalse(MODULE_1.equals(null));
+
+        // different module -> returns false
+        assertFalse(MODULE_1.equals(MODULE_2));
+
+        // same name, different tasks -> returns true
+        List<Task> taskList = TypicalTasks.getTypicalTasksForModule(MODULE_NAME_0);
+        Module differentTask = new ModuleBuilder()
+                .withName(MODULE_NAME_0)
+                .withStudents(TypicalStudents.getTypicalStudents())
+                //.withTasks(taskList) // this line will cause Duplicate Task Exception for some reason
```

Comment:
I think we should either fix the issue or just remove this line entirely

Reply:
removed

Reply:
I think the amy and bob objects get reused, thus causing the issue

Reply:
One solution will be to create new amy and bob object each time the getTypicalStudents is called to avoid using the same variable repeatedly.

```
 public static List<Student> getTypicalStudents() {
        final Student amy = new StudentBuilder().withName(VALID_NAME_AMY).withStudentId(VALID_STUDENT_ID_AMY)
                .withTeleHandle(VALID_TELE_HANDLE_AMY).withEmail(VALID_EMAIL_AMY).build();
        final Student bob = new StudentBuilder().withName(VALID_NAME_BOB).withStudentId(VALID_STUDENT_ID_BOB)
                .withTeleHandle(VALID_TELE_HANDLE_BOB).withEmail(VALID_EMAIL_BOB).build();
        return new ArrayList<>(Arrays.asList(amy, bob));
    }
```

Reply:
Ok upon looking at the original approach in AB3  to this is still using the static variable. But each time any modification is made, it is made on a copy of the original object e.g. Amy's copy so the original static final variable won't get modified.

Reply:
Edited `getTypicalStudents` and add back the commented out lines in the latest commits. Can take a look. @Ruofan1023 @tanjoewel